### PR TITLE
win_command: Correctly handle command arguments

### LINKED ISSUE #23019: win_command incorrectly parses args containing spaces
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME
win_command

##### ANSIBLE VERSION

```
ansible 2.2.1.0
```

##### CONFIGURATION
n/a

##### OS / ENVIRONMENT
Windows Server 2012 r2 (target host); Ubuntu 16.04 (management system)

##### SUMMARY
Double-quoted commands containing spaces are incorrectly parsed by "win_command.ps1", causing arguments to be split incorrectly before execution.

##### STEPS TO REPRODUCE

Example playbook demonstrating this issue:

```yaml

- hosts: windows
  gather_facts: false
  tasks:
  - win_command: C:\windows\system32\netsh.exe advfirewall firewall show rule name="File and Printer Sharing (SMB-In)"
    register: win_command_output
    failed_when: false
  - debug: var=win_command_output
  vars:
    ansible_user: Administrator
    ansible_password: FIXME
    ansible_port: 5986
    ansible_connection: winrm
    ansible_winrm_server_cert_validation: ignore
```

##### EXPECTED RESULTS

* win_command_output.rc == 0
* win_command_output.stdout encodes a table of firewall rule information.

##### ACTUAL RESULTS
* win_command_output.rc == 1
* win_command_output.stdout contains an usage/invocation error

##### NOTES
* The issue can be traced to incorrect handling of the command string _after_ it is passed to "win_command.ps1".
* The _relevant_ processing logic of the current implementation can be summarized as follows:
  1. It parses the command from a single string into a list of strings (and appears to do so correctly).
  1. It uses that list to infer the "command" (ie list's first entry), as distinct from the "arguments" (ie everything in the list _after_ the first entry, if any).
  1. It then re-assembles a single string containing _just_ the arguments by simply joining the individual arguments strings together with a space between each entry. **This is incorrect behavior whenever there are any arguments containing a space, tab, or double-quote.**
* For MS' escaping rules, see https://msdn.microsoft.com/en-us/library/17w5ykft.aspx.
* Workarounds may be developed using either "win_raw" or "win_shell".

### PR DESCRIPTION
**Pull request by dagwieers**
##### SUMMARY

This fixes #23019

##### ISSUE TYPE

 - Bugfix Pull Request

##### COMPONENT NAME

win_command

##### ANSIBLE VERSION

v2.3

### CODE REVIEW

**Comment by nitzmahone** [comment]
If I read this right, you just turned `win_command` into `win_shell`- IIRC the only reason that `command` *has* `executable` is because it shares its impl with `shell`.

**Comment by nitzmahone** [comment]
There shouldn't be an executable- if anything, we'd just use `CreateProcess` directly...

**Comment by nitzmahone** [comment]
Can you describe an actual thing that this fixes?

**Comment by dagwieers** [comment]
We don't just concat all the arguments together in a large string, when the arguments should be the original array. (If you read the issue, you'd know).

I added executable because:
- the command module allows the same thing, and it is very convenient (rather than doing `cmd /c "foo bar"`)
- it is now using powershell to run a command, whereas I would have expected it using cmd.exe instead

The original implementation didn't default executable to `cmd.exe`.

**Comment by nitzmahone** [comment]
Ah, I missed the bug report. Yeah, this isn't the way to fix that- `command` is basically "run a process", so `win_command` should remain the analogue to that, not "use cmd"- that'd be `win_shell` with executable == `cmd.exe`. I delegated the command line parsing to the Win32 `CommandLineToArgvW` API function (which IIRC is exactly what cmd.exe does, too), so it's possible that there's something else that's breaking the parsing that we might want to fix.

**Comment by dagwieers** [comment]
@nitzmahone This is the fix: https://github.com/ansible/ansible/pull/23055/files#diff-2e1e53e798f59df98c4fd5a175af0c3eR127

**Comment by dagwieers** [comment]
Still, I think it is preferable to add `executable: cmd` rather than wrap everything in `cmd /c "foo bar"` which makes it harder to quote correctly.

**Comment by nitzmahone** [comment]
But if you want to exec under a shell, use `win_shell`- it already does all that.

The actual fix (sans all the executable stuff) looks fine, but IIRC there's a reason I didn't do it that way. It might be that I was having trouble with pathing or something- I ran into that with the exec wrapper (where `CreateProcess` doesn't always properly find a naked executable, so had to manually call `SearchPath` to find it first). If that's working, though, I'm happy to take that change alone for 2.3.

**Comment by dagwieers** [comment]
@nitzmahone I don't see why the `executable` option would be documented in the command module, if it would not be working. That would be a bug in itself then. My interpretation is that it is on purpose so you can run e.g. expect as executable and provide the expect script/commands.

**Comment by nitzmahone** [comment]
It's not that it doesn't work in `command`, but talking with @jimi-c and @bcoca, they tend to agree that it's likely a nonsensical accident of history that it's defined on `command` at all (due to shared code). @bcoca even proposed deprecating it and removing from docs. 

I really don't want to add `executable` to command, but if you want to cut this down to just the parsing change, I'm happy to take that.

**Comment by dagwieers** [comment]
Ok, fair enough. I agree that it should be removed from the command module documentation too.

**Comment by nitzmahone** [comment]
PRs welcome ;)

**Comment by dagwieers** [comment]
So, I reverted everything related to the executable parameter. However, there are a few cleanups related to win_command/win_shell consistency and docs. You want that in another PR, or not having it at all ?

**Comment by nitzmahone** [comment]
s'ok

File: lib/ansible/modules/windows/win_command.ps1 [hunk 1]
```diff
@@ -117,19 +117,21 @@ $exec_args = $null
 # FUTURE: no longer necessary once we switch to raw Win32 CreateProcess
 $parsed_command_line = [Ansible.Command.NativeUtil]::ParseCommandLine($raw_command_line);
 $exec_application = $parsed_command_line[0]
-If($parsed_command_line.Length -gt 1) {
-    # lop the application off, then rejoin the args as a single string
-    $exec_args = $parsed_command_line[1..$($parsed_command_line.Length-1)] -join " "
+If ($parsed_command_line.Length -gt 1) {
+    $exec_args = $parsed_command_line[1..-1]
```
Comment (nitzmahone) [inline lib/ansible/modules/windows/win_command.ps1]: PS slices aren't the same as Python (I *hate* the way the PS ones work)- this is incorrect, it needs to be the way I originally had it. Try the following: ` @(1,2,3,4)[1..-1]`, you'll see that it reorders the args and picks up the wrong ones.

**Review by nitzmahone — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/windows/win_command.ps1 (hunk 1 above)
Comment (dagwieers) [reply lib/ansible/modules/windows/win_command.ps1]: What ? That makes no sense :-)
It showed 2 1 4 in my case.
What's the purpose of this behaviour ?

PS I looked it up and stackexchange told me to do it like that ;-)

Comment (nitzmahone) [reply lib/ansible/modules/windows/win_command.ps1]: (`1..-1` creates a range of values `@(1,0,-1)`)

Comment (nitzmahone) [reply lib/ansible/modules/windows/win_command.ps1]: PS doesn't really have array slices, so go downvote that SO answer ;)- it accepts a list of indices in the brackets, which the range operator `..` can be used to generate in limited cases.

**Comment by dagwieers** [comment]
@nitzmahone Shippable errors are still problematic, but unrelated to the PR...

**Comment by nitzmahone** [comment]
It doesn't appear that this fixes the referenced bug (I tried it- doesn't work with this change in place)- I think running it through cmd.exe might have, just because the shell reparsed the command line, but your change is effectively no change (since assigning the `exec_args` array to `$psi.Arguments` just turns it into a space-separated string anyway). `Arguments` is a scalar string (see [here](https://msdn.microsoft.com/en-us/library/system.diagnostics.processstartinfo.arguments(v=vs.110).aspx))...

**Comment by dagwieers** [comment]
Sigh. Again not how Unix works with arguments I guess :-/

**Comment by nitzmahone** [comment]
Actually, I just dumped the raw arg values- the quotes never made it. So it's probably a YAML issue (or something in between).

**Comment by nitzmahone** [comment]
Hmm, it's correct up to the _raw_params that are getting sent to the module...

**Comment by nitzmahone** [comment]
Dammit- it's Microsoft's own function that's eating the quotes (`CommandLineToArgvW`).

**Comment by nitzmahone** [comment]
Gonna zap this one, since the fix will be more involved (and probably wait for 2.4, since we can workaround with win_shell for now).

**Pull request closed without merging**